src/transformers/models/idefics/modeling_idefics.py

```diff
@@ -1435,6 +1461,98 @@ def forward(
+                        f"You passed `inputs_embeds` to `.generate()`, but the model class {self.__class__.__name__} "
+                        "doesn't have its forwarding implemented. See the GPT2 implementation for an example "
+                        "(https://github.com/huggingface/transformers/pull/21405), and feel free to open a PR with it!"
+                    )
+                # In this case, `input_ids` is moved to the `model_kwargs`, so a few automations (like the creation of
+                # the attention mask) can rely on the actual model input.
+                model_kwargs["input_ids"] = self._maybe_initialize_input_ids_for_generation(
+                    inputs, bos_token_id, model_kwargs=model_kwargs
+                )
+            else:
+                if inputs is not None:
+                    raise ValueError("You passed `inputs_embeds` and `input_ids` to `.generate()`. Please pick one.")
+            inputs, input_name = model_kwargs["inputs_embeds"], "inputs_embeds"
+
+        # 4. if `inputs` is still None, try to create `input_ids` from BOS token
+        inputs = self._maybe_initialize_input_ids_for_generation(inputs, bos_token_id, model_kwargs)
+
+        # 5. Prepare model_kwargs for IDEFICS vision component
+        pixel_values = model_kwargs.get("pixel_values", None)
+        image_encoder_embeddings = model_kwargs.get("image_encoder_embeddings", None)
+        perceiver_embeddings = model_kwargs.get("perceiver_embeddings", None)
+
+        if pixel_values is not None:
+            batch_size, num_images = pixel_values.shape[:2]
+            pixel_values = pixel_values.contiguous().view(batch_size * num_images, *pixel_values.shape[2:])
+            image_encoder_embeddings = self.model.vision_model(pixel_values=pixel_values).last_hidden_state
+
+        elif image_encoder_embeddings is not None:
+            batch_size, num_images, image_seq_len, image_hidden_size = image_encoder_embeddings.size()
+            image_encoder_embeddings = image_encoder_embeddings.view(
+                batch_size * num_images, image_seq_len, image_hidden_size
+            )
+
+        if self.config.use_resampler:
+            if perceiver_embeddings is None:
+                perceiver_embeddings = self.model.perceiver_resampler(image_encoder_embeddings)
+                image_seq_len, image_hidden_size = perceiver_embeddings.size(1), perceiver_embeddings.size(2)
+                model_kwargs["perceiver_embeddings"] = perceiver_embeddings.view(
+                    batch_size, num_images, image_seq_len, image_hidden_size
+                )
+            else:
+                model_kwargs["perceiver_embeddings"] = perceiver_embeddings
+        else:
+            image_seq_len, image_hidden_size = image_encoder_embeddings.size(1), image_encoder_embeddings.size(2)
+            model_kwargs["image_encoder_embeddings"] = image_encoder_embeddings.view(
+                batch_size, num_images, image_seq_len, image_hidden_size
```

Comment:
Regarding @VictorSanh's comment, it makes more sense indeed to compute these values in the forward, and return them as outputs. Our API usually works this way so let's keep this. It should also simplify this function!

Reply:
Okk will look into this